core/internal/runhistoryreader/reader.go

```diff
@@ -59,6 +60,41 @@ func New(
 		return nil, err
 	}
 
+	partitions := make([]iterator.RowIterator, len(historyReader.parquetFiles))
+	for i, parquetFile := range historyReader.parquetFiles {
+		selectedRows := iterator.SelectRows(
+			parquetFile,
+			iterator.StepKey,
+			0,
+			float64(math.MaxInt64),
+			false,
+		)
+		selectedColumns, err := iterator.SelectColumns(
+			iterator.StepKey,
+			historyReader.keys,
+			parquetFile.ParquetReader().MetaData().Schema,
+			len(historyReader.keys) == 0,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		rowIterator, err := iterator.NewRowIterator(
+			ctx,
+			parquetFile,
+			selectedRows,
+			selectedColumns,
+		)
+		if err != nil {
+			for _, partition := range partitions {
+				partition.Release()
+			}
```

Comment:
`Release` is only called here and not in the other error check?

This is a tricky pattern. It's more robust to use a defer statement with named return values to ensure you only `Release` on error. You should also extract this entire thing into a helper function; it's too long for the constructor.

```go
partitions := make([]iterator.RowIterator, 0, <capacity>)
defer func() {
	if err != nil {
		for _, partition := range partitions {
			partition.Release()
		}
	}
}()
```

Also, isn't this a multi iterator? `NewMultiIterator(partitions).Release()` would work, although it doesn't feel like it's designed for that.